### PR TITLE
Streamline the build process

### PR DESCRIPTION
**Pull request by Ornias1993**
**Description of the change**

This PR reworks the build system to include tests on both PR and Push.
It also makes sure github caches are used slightly more efficiently.

It does 3 things:
- Test building all platforms (using cache if available), fail if the one of them fails, cache the results, don't push
- Build local test build (using cache if available), if GOSS tests are available, don't push
- Execute Gosstest using previous local build, if GOSS tests are available,, fail if test fails
- Build and Push the container (using cache if available)

**Benefits**

With this included we prevent pushing broken updates and CI speed should increase slightly.

**Possible drawbacks**

Not that I'm aware of.

**Applicable issues**

- fixes #

**Additional information**

Test PR is available here:
https://github.com/truecharts/containers/pull/1/checks?check_run_id=2106318952

### CODE REVIEW

File: .github/workflows/build.yaml [hunk 1]
```diff
@@ -88,8 +97,7 @@ jobs:
         restore-keys: |
           ${{ runner.os }}-buildx-
 
-    - name: Build and Push (PR)
-      if: github.event_name == 'pull_request'
+    - name: Try Build
```
Comment (angelnu) [inline .github/workflows/build.yaml:100]: I suggest to remove this section completely - the build at the bottom should cover all builds and the push for non-PRs

File: .github/workflows/build.yaml [hunk 2]
```diff
@@ -52,6 +52,15 @@ jobs:
         else
           echo ::set-output name=goss::false
         fi
+        if [ "${{github.event_name}}" == "pull_request" ]; then
+          buildURL1="${{ matrix.container }}:test"
+          buildURL2="pull-request:${{github.event.number}}"
+        else
+          buildURL1="${{ matrix.container }}:latest"
```
Comment (angelnu) [inline .github/workflows/build.yaml:59]: I think we can always use there - for PRs anyway we cannot push

File: .github/workflows/build.yaml [hunk 3]
```diff
@@ -132,5 +144,6 @@ jobs:
         file: ./${{ matrix.container }}/Dockerfile
         push: true
```
Comment (angelnu) [inline .github/workflows/build.yaml:145]: here we would need this true when no PR

File: .github/workflows/build.yaml [hunk 4]
```diff
@@ -98,31 +106,35 @@ jobs:
...
         GOSS_FILE: ${{ matrix.container }}/goss.yaml
       run: |
-        dgoss run ghcr.io/${{ github.repository_owner }}/pull-request:${{github.event.number}}
+        dgoss run ghcr.io/${{ github.repository_owner }}/${{ steps.prep.outputs.buildURL2 }}
+
 
-    - name: Build and Push (non PR)
+
+    - name: Build and Push
       if: github.event_name != 'pull_request'
```
Comment (angelnu) [inline .github/workflows/build.yaml:138]: if this is the same common build task then it should always be done

**Review by angelnu — CHANGES_REQUESTED** [review]
good progress, looking forward for a few extra changes.

File: .github/workflows/build.yaml (hunk 2 above)
Comment (Ornias1993) [reply .github/workflows/build.yaml:59]: You mean always use the production tags (not PR numbered tags) even for non-push builds?

**Review by onedr0p — APPROVED** [review]

**Pull request merged**